File: api/khairat-api/src/main/java/my/emasjid/khairatapi/service/MemberService.java
package my.emasjid.khairatapi.service;

import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import my.emasjid.khairatapi.entity.Dependent;
import my.emasjid.khairatapi.entity.Member;
import my.emasjid.khairatapi.entity.MemberTag;
import my.emasjid.khairatapi.entity.PaymentHistory;
import my.emasjid.khairatapi.entity.Person;
import my.emasjid.khairatapi.repository.DependentRepository;
import my.emasjid.khairatapi.repository.MemberRepository;
import my.emasjid.khairatapi.repository.MemberTagRepository;
import my.emasjid.khairatapi.repository.PaymentHistoryRepository;
import my.emasjid.khairatapi.repository.PersonRepository;


@Service
public class MemberService {

    Logger logger = LoggerFactory.getLogger(MemberService.class);
    @Autowired
    private MemberRepository memberRepository;
    @Autowired
    private PersonRepository personRepository;
    @Autowired
    private DependentRepository dependentRepository;
    @Autowired
    private MemberTagRepository memberTagRepository;
    @Autowired
    private PaymentHistoryRepository paymentHistoryRepository;
    @Autowired
    private PaymentService paymentService;

    public Member save(Member member) {
        Long memberId = member.getId();
        List<PaymentHistory> paymentHistory = member.getPaymentHistories();

        if (memberId == null) {
            Person person = member.getPerson();
            List<MemberTag> memberTags = member.getMemberTags();
            List<Dependent> dependents = member.getDependents();
    
            person = personRepository.save(person);
    
            member.setPerson(person);
            member.setMemberTags(null);
            member.setDependents(null);
            member = memberRepository.save(member);
    
            final Member newMember = member;
    
            memberTags.forEach(tags -> {
                tags.setMember(newMember);
            });
    
            dependents.forEach(dependent -> {
                Person personD = dependent.getPerson();
                personD = personRepository.save(personD);
                dependent.setPerson(personD);
                dependent.setMember(newMember);
            });

            paymentHistory.forEach(payment -> {
                payment.setMember(newMember);
            });
    
            memberTags = memberTagRepository.saveAll(memberTags);
            dependents = dependentRepository.saveAll(dependents);
            paymentHistory = paymentHistoryRepository.saveAll(paymentHistory);

        } else {
            Person person = member.getPerson();
            List<MemberTag> memberTags = member.getMemberTags();
            // List<Dependent> dependents = member.getDependents();

            person = personRepository.save(person);
            member.setPerson(person);
            member.setMemberTags(null);
            member.setDependents(null);
    
            final Member newMember = member;

            memberTagRepository.deleteByMemberId(memberId);
            memberTags.forEach(tags -> {
                tags.setMember(newMember);
            });
    
            // dependentRepository.deleteByMemberId(memberId);
            // dependents.forEach(dependent -> {
            //     Person personD = dependent.getPerson();
            //     if(personD.getId() != null){
            //         personRepository.delete(personD);
            //     }
            // });
            // dependents.forEach(dependent -> {
            //     Person personD = dependent.getPerson();
            //     if(personD.getId() == null){
            //         personD = personRepository.save(personD);
            //     }
            //     dependent.setPerson(personD);
            //     dependent.setMember(newMember);
            // });
            // dependents = dependentRepository.saveAll(dependents);
    
            memberTags = memberTagRepository.saveAll(memberTags);
            updatePaymentHistory(paymentHistory, memberId);
        }

        return member;
    }

private void updatePaymentHistory(List<PaymentHistory> paymentHistory, Long memberId) {
    if (paymentHistory.isEmpty()) {
        paymentService.deleteCurrentYearPaymentHistory(memberId);
    } else {
        if (!paymentService.isCurrentYearPaymentExist(memberId)) {
            for (PaymentHistory history : paymentHistory) {
                if (history.getId() == null) {
                    Member member = findOne(memberId);
                    history.setMember(member);
                    paymentService.save(history);
                    break;
                }
            }
        }
    }
}

    public Member findOne(Long id) {
        return memberRepository.findById(id).get();
    }

    public Page<Member> findAll(Pageable pageable) {
        return memberRepository.findAll(pageable);
    }

    public List<Member> findBy(Person person){
        return memberRepository.findByPersonNameContainingIgnoreCaseOrPersonIcNumberContainingIgnoreCaseOrPersonPhoneContainingIgnoreCaseOrPersonAddressContainingIgnoreCase(
            person.getName(), person.getIcNumber(), person.getPhone(), person.getAddress());
    }

    public List<Member> findAllByOrderByPersonName(){
        return memberRepository.findAllByOrderByPersonName();
    }

    public Long count(){
        return memberRepository.count();
    }

    // public Double paidPercentage(){
    //     return memberRepository.paidPercentage();
    // }
}

File: api/khairat-api-go/service/member_service.go
package service

import (
	"github.com/Dev4w4n/e-masjid.my/api/khairat-api/model"
	"github.com/Dev4w4n/e-masjid.my/api/khairat-api/repository"
)

type MemberService interface {
	Save(member model.Member) (model.Member, error)
	FindOne(id int) (model.Member, error)
	FindAll() (model.Response, error)
	FindByQuery(query string) ([]model.Member, error)
	FindAllOrderbyPersonName() ([]model.Member, error)
	CountAll() (int64, error)
	FindAllByTagIdOrderByMemberName(tagIdStr string) ([]model.Member, error)
}

type MemberServiceImpl struct {
	memberRepository         repository.MemberRepository
	personRepository         repository.PersonRepository
	dependentRepository      repository.DependentRepository
	memberTagRepository      repository.MemberTagRepository
	paymentHistoryRepository repository.PaymentHistoryRepository
}

func NewMemberService(memberRepository repository.MemberRepository,
	personRepository repository.PersonRepository,
	dependentRepository repository.DependentRepository,
	memberTagRepository repository.MemberTagRepository,
	paymentHistoryRepository repository.PaymentHistoryRepository) MemberService {
	return &MemberServiceImpl{
		memberRepository:         memberRepository,
		personRepository:         personRepository,
		dependentRepository:      dependentRepository,
		memberTagRepository:      memberTagRepository,
		paymentHistoryRepository: paymentHistoryRepository,
	}
}

// CountAll implements MemberService.
func (repo *MemberServiceImpl) CountAll() (int64, error) {
	result, err := repo.memberRepository.CountAll()

	if err != nil {
		return 0, err
	}

	return result, nil
}

// FindAll implements MemberService.
func (repo *MemberServiceImpl) FindAll() (model.Response, error) {
	var response model.Response

	result, err := repo.memberRepository.FindAll()

	if err != nil {
		return response, err
	}

	response.Content = result

	return response, nil
}

// FindAllByTagIdOrderByMemberName implements MemberService.
func (repo *MemberServiceImpl) FindAllByTagIdOrderByMemberName(tagIdStr string) ([]model.Member, error) {
	var result []model.Member

	result, err := repo.memberRepository.FindByTagOrderByMemberNameAsc(tagIdStr)

	if err != nil {
		return nil, err
	}

	return result, nil
}

// FindAllOrderbyPersonName implements MemberService.
func (repo *MemberServiceImpl) FindAllOrderbyPersonName() ([]model.Member, error) {
	var result []model.Member

	result, err := repo.memberRepository.FindAllOrderByPersonName()

	if err != nil {
		return nil, err
	}

	return result, nil
}

// FindBy implements MemberService.
func (repo *MemberServiceImpl) FindByQuery(query string) ([]model.Member, error) {
	var result []model.Member

	result, err := repo.memberRepository.FindByQuery(query)

	if err != nil {
		return nil, err
	}

	return result, nil
}

// FindOne implements MemberService.
func (repo *MemberServiceImpl) FindOne(id int) (model.Member, error) {
	var result model.Member

	result, err := repo.memberRepository.FindById(id)

	if err != nil {
		return result, err
	}

	return result, nil
}

// Save implements MemberService.
func (repo *MemberServiceImpl) Save(member model.Member) (model.Member, error) {

	if member.Id == 0 {
		payments := member.PaymentHistory
		dependents := member.Dependents
		tags := member.MemberTags
		newPerson, err := repo.personRepository.Save(member.Person)

		if err != nil {
			return model.Member{}, err
		}

		member.PersonId = newPerson.Id
		member.Dependents = nil
		member.MemberTags = nil
		member.PaymentHistory = nil

		member, err := repo.memberRepository.Save(member)

		if err != nil {
			return model.Member{}, err
		}

		for _, dependent := range dependents {
			person := dependent.Person
			newPerson, err := repo.personRepository.Save(person)

			if err != nil {
				return model.Member{}, err
			}

			dependent.MemberId = member.Id
			dependent.PersonId = newPerson.Id

			err = repo.dependentRepository.Save(dependent, member.Id)

			if err != nil {
				return model.Member{}, err
			}
		}

		for _, tag := range tags {
			tag.MemberId = member.Id

			_, err = repo.memberTagRepository.Save(tag)

			if err != nil {
				return model.Member{}, err
			}
		}

		for _, payment := range payments {
			payment.MemberId = member.Id

			err = repo.paymentHistoryRepository.Save(payment)

			if err != nil {
				return model.Member{}, err
			}
		}

	} else {
		person := member.Person
		memberTags := member.MemberTags

		_, err := repo.personRepository.Save(person)

		if err != nil {
			return model.Member{}, err
		}

		err = repo.memberTagRepository.DeleteByMemberId(member.Id)

		if err != nil {
			return model.Member{}, err
		}

		for _, tag := range memberTags {
			tag.MemberId = member.Id

			_, err = repo.memberTagRepository.Save(tag)

			if err != nil {
				return model.Member{}, err
			}
		}

		err = repo.paymentHistoryRepository.UpdatePaymentHistory(member.PaymentHistory,
			member.Id)

		if err != nil {
			return model.Member{}, err
		}
	}

	return member, nil
}
